opentelemetry-sdk/src/logs/log_emitter.rs

```diff
@@ -97,25 +100,46 @@ impl LoggerProvider {
     }
 
     /// Force flush all remaining logs in log processors and return results.
-    pub fn force_flush(&self) -> Vec<LogResult<()>> {
-        self.log_processors()
-            .iter()
-            .map(|processor| processor.force_flush())
-            .collect()
+    pub fn force_flush(&self) -> LogResult<()> {
+        // propagate force_flush to processors
+        let mut errs = vec![];
+        for processor in &self.inner.processors {
+            if let Err(err) = processor.force_flush() {
+                errs.push(err);
+            }
+        }
+
+        if errs.is_empty() {
+            Ok(())
+        } else {
+            Err(LogError::Other(format!("{errs:?}").into()))
+        }
     }
 
     /// Shuts down this `LoggerProvider`
-    pub fn shutdown(&self) -> Vec<LogResult<()>> {
-        // mark itself as already shutdown
-        self.is_shutdown
-            .store(true, std::sync::atomic::Ordering::Relaxed);
-        // propagate the shutdown signal to processors
-        // it's up to the processor to properly block new logs after shutdown
-        self.inner
-            .processors
-            .iter()
-            .map(|processor| processor.shutdown())
-            .collect()
+    pub fn shutdown(&self) -> LogResult<()> {
+        if self
+            .is_shutdown
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+            .is_ok()
+        {
+            // propagate the shutdown signal to processors
+            // it's up to the processor to properly block new logs after shutdown
+            let mut errs = vec![];
+            for processor in &self.inner.processors {
+                if let Err(err) = processor.shutdown() {
+                    errs.push(err);
+                }
+            }
+
+            if errs.is_empty() {
+                Ok(())
+            } else {
+                Err(LogError::Other(format!("{errs:?}").into()))
+            }
+        } else {
+            Err(LogError::Other("logger provider already shut down".into()))
```

Comment:
We should make a dedicate LogError variant for repeat shutdown.

Reply:
Agree to fix it. Not in this PR, as the current structure of placing these errors in the API is not correct, and needs a refactoring. Will cover it as part of https://github.com/open-telemetry/opentelemetry-rust/issues/1042

Reply:
I'd mark this as releasing blocking and add a todo in code too. It will be very anti-intutive if the users has to parse the error string to understand if the errors is caused by repeated shutdown

Reply:
> I'd mark this as releasing blocking

yes for stable or even beta release. But not for https://github.com/open-telemetry/opentelemetry-rust/pull/1738 which is still alpha for logs.

Reply:
Sorry if this is already answered. What is wrong with returning vec of LogResult for shutdown? Is it only for consistency with metrics?

Reply:
yes. And also, if any individual sub-component faced issue while shutdown, they should be emitting their own logging. Shutdown() just need to capture the overall state. But I am not very sure we are doing it correctly. The proper solution should be coming once we address https://github.com/open-telemetry/opentelemetry-rust/issues/1146 (that also is very much related to https://github.com/open-telemetry/opentelemetry-rust/issues/761 as it is hard to solve internal logging without fixing the infinite telemetry/circular dep problem)

Reply:
Ok thanks, that makes sense.